tests/008-ed-keys.cpp

```diff
@@ -0,0 +1,147 @@
+        };
+    };
+
+    TEST_CASE("008 - Ed Keys: Connections", "[001][keys][tls][connection]")
+    {
+
+        auto client_seed = oxenc::from_hex("468e7ed2cd914ca44568e7189245c7b8e5488404fc88a4019c73b51d9dbc48a5");
+        auto client_pubkey = oxenc::from_hex("626136fe40c8860ee5bdc57fd9f15a03ef6777bb9237c18fc4d7ef2aacfe4f88");
+        oxen::quic::gnutls_key client_pubkey_array;
+        std::copy(client_pubkey.begin(), client_pubkey.end(), client_pubkey_array.data());
+
+        auto server_seed = oxenc::from_hex("fefbb50cdd4cde3be0ae75042c44ff42b026def4fd6be4fb1dc6e81ea0480c9b");
+        auto server_pubkey = oxenc::from_hex("d580d5c68937095ea997f6a88f07a86cdd26dfa0d7d268e80ea9bbb5f3ca0304");
+        oxen::quic::gnutls_key server_pubkey_array;
+        std::copy(server_pubkey.begin(), server_pubkey.end(), server_pubkey_array.data());
+
+        Network test_net{};
+
+        auto client_tls = GNUTLSCreds::make_from_ed_keys(client_seed, client_pubkey);
+        auto server_tls = GNUTLSCreds::make_from_ed_keys(server_seed, server_pubkey, /*snode = */ true);
+        auto server_tls2 = GNUTLSCreds::make_from_ed_keys(client_seed, client_pubkey, /*snode = */ true);
+
+        opt::local_addr server_local{};
+        opt::local_addr client_local{};
+
+        auto server_endpoint = test_net.endpoint(server_local);
+        REQUIRE(server_endpoint->listen(server_tls));
+
+        opt::local_addr client_remote{"127.0.0.1"s, server_endpoint->local().port()};
+
+        auto client_endpoint = test_net.endpoint(client_local);
+
+        std::promise<bool> key_was_allowed;
+        auto f = key_was_allowed.get_future();
+
+        auto client_key_allowed_cb = [&](const oxen::quic::gnutls_key& key, bool is_relay) {
+            if (not is_relay)
+            {
+                key_was_allowed.set_value(true);
+                return true;
+            }
+
+            if (key == client_pubkey_array)
+            {
+                key_was_allowed.set_value(true);
+                return true;
+            }
+
+            log::error(
+                    log_cat,
+                    "Key mismatch:\n{}\n{}",
+                    oxenc::to_hex(key.begin(), key.end()),
+                    oxenc::to_hex(client_pubkey_array.begin(), client_pubkey_array.end()));
+
+            key_was_allowed.set_value(false);
+            return false;
+        };
+
+        client_tls->set_key_verify_callback(client_key_allowed_cb);
+        server_tls2->set_key_verify_callback([&](auto, auto) {
+            key_was_allowed.set_value(true);
+            return true;
+        });
+
+        SECTION("Successful Connection")
+        {
+            server_tls->set_key_verify_callback([&](const auto& key, bool is_relay) {
+                key_was_allowed.set_value(true);
+                return true;
+            });
+
+            REQUIRE_NOTHROW(client_endpoint->connect(client_remote, client_tls));
+
+            REQUIRE(f.wait_for(20ms) == std::future_status::ready);
```

Comment:
Are these sleeps being used to give the client time to establish a connection? You could use the same logic in the other tests like [001](https://github.com/oxen-io/oxen-libquic/blob/857188c5d21ea01819c0d88258b84f592ed05125/tests/001-handshake.cpp#L190), where the `client_tls_policy` is given a callback that signals "connection ready"

Reply:
This is a future wait with a timeout, rather than a sleep.  It's so the test fails rather than hanging indefinitely if something goes horribly wrong.

Reply:
> Are these sleeps

Not a sleep -- this just puts an upper limit on how long it can wait before returning.

> We don't do this in any other test, and any failure case where it hangs indefinitely should be handled internally

We probably should -- if a failure happens in a test now the test suite hangs indefinitely, rather than failing, where failing would be nicer.

Perhaps putting this in a common test suite function that we can use in many places, something like:
```c++
template <typename F>
void require_future(F& f, std::chrono::milliseconds timeout = 1s) { REQUIRE(f.wait_for(timeout) == std::future_status::ready); }
```
That said, 20ms seems way too short for this (because this is just a fail-safe in case it hangs).

Reply:
> This is a future wait with a timeout, rather than a sleep

> Not a sleep 

I was using "sleep" in the notional rather than literal sense

My main issue with this is using certificate verification as a proxy for "connection ready." We know already that this is not how ngtcp2 signals that at all